protocol/messenger.go

```diff
@@ -1750,3 +1919,693 @@ func (m *Messenger) VerifyENSNames(rpcEndpoint, contractAddress string, ensDetai
+	message.Clock = clock
+	message.Timestamp = timestamp
+	message.Text = "Transaction sent"
+
+	request := &protobuf.SendTransaction{
+		Clock:           message.Clock,
+		TransactionHash: transactionHash,
+		Signature:       signature,
+	}
+	encodedMessage, err := proto.Marshal(request)
+	if err != nil {
+		return nil, err
+	}
+
+	newMessageID, err := m.dispatchMessage(ctx, &RawMessage{
+		LocalChatID:         chat.ID,
+		Payload:             encodedMessage,
+		MessageType:         protobuf.ApplicationMetadataMessage_SEND_TRANSACTION,
+		ResendAutomatically: true,
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	message.ID = types.EncodeHex(newMessageID)
+	message.CommandParameters = &CommandParameters{
+		TransactionHash: transactionHash,
+		Signature:       signature,
+		CommandState:    CommandStateTransactionSent,
+	}
+
+	err = message.PrepareContent()
+	if err != nil {
+		return nil, err
+	}
+
+	err = chat.UpdateFromMessage(message)
+	if err != nil {
+		return nil, err
+	}
+
+	err = m.persistence.SaveMessagesLegacy([]*Message{message})
+	if err != nil {
+		return nil, err
+	}
+
+	response.Chats = []*Chat{chat}
+	response.Messages = []*Message{message}
+	return &response, m.saveChat(chat)
+}
+
+func (m *Messenger) ValidateTransactions(client EthClient, addresses []types.Address) (*MessengerResponse, error) {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	modifiedChats := make(map[string]bool)
+
+	logger := m.logger.With(zap.String("site", "ValidateTransactions"))
+	logger.Debug("Validating transactions")
+	txs, err := m.persistence.TransactionsToValidate()
+	if err != nil {
+		logger.Error("Error pulling", zap.Error(err))
+		return nil, err
+	}
+	logger.Debug("Txs", zap.Int("count", len(txs)), zap.Any("txs", txs))
+	var response MessengerResponse
+	validator := NewTransactionValidator(addresses, m.persistence, m.logger)
+	responses, err := validator.ValidateTransactions(client)
+	if err != nil {
+		logger.Error("Error validating", zap.Error(err))
+		return nil, err
+	}
+	for _, validationResult := range responses {
+		var message *Message
+		var chat *Chat
+		var ok bool
+		chatID := types.EncodeHex(crypto.FromECDSAPub(validationResult.Transaction.From))
+		chat, ok = m.allChats[chatID]
+		if !ok {
+			chat = OneToOneFromPublicKey(validationResult.Transaction.From)
+		}
+		if validationResult.Message != nil {
+			message = validationResult.Message
+		} else {
+			message = &Message{}
+			err := extendMessageFromChat(message, chat, &m.identity.PublicKey)
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		message.MessageType = protobuf.ChatMessage_ONE_TO_ONE
+		message.ContentType = protobuf.ChatMessage_TRANSACTION_COMMAND
+		message.LocalChatID = chatID
+		message.OutgoingStatus = ""
+
+		clock, timestamp := chat.NextClockAndTimestamp()
+		message.Clock = clock
+		message.Timestamp = timestamp
+		message.Text = "Transaction received"
+
+		message.ID = validationResult.Transaction.MessageID
```

Comment:
Maybe there can be something like `extendMessageFromTxValidation()` similar to `extendMessageFromChat`?

Reply:
I'd wait a bit to make sure implementation settles before refactoring, as this might still have issues/request for changes etc, if that's ok